File: CombatTracker-Application/CombatTrackerData/CreatureCharts/LevelChart.vb
Partial Class LevelChart
    Inherits IEntity(Of LevelChart)
    Public Overrides Function ToString() As String
        Return Name
    End Function

    Public Shared Function LoadByName(ByVal DB As RMSSDataDataContext, ByVal Name As String) As LevelChart
        Return (From SC In DB.LevelCharts Where SC.Name = Name Select SC).FirstOrDefault
    End Function

    Public Function LookupLevel(ByVal Roll As Integer) As Integer
        Return (From LC In Me.LevelChartValues Where LC.Min <= Roll AndAlso Roll <= LC.Max Select LC.Mod).FirstOrDefault
    End Function



    Private Sub OnValidate(ByVal action As System.Data.Linq.ChangeAction)
        If action <> Data.Linq.ChangeAction.Delete AndAlso Not IsValid Then
            Throw New ApplicationException("Rule violations prevent saving")
        End If
    End Sub

    Public Overrides Function GetRuleViolations() As System.Collections.Generic.List(Of RuleViolation)
        Dim L As New List(Of RuleViolation)
        'If String.IsNullOrEmpty(Name) Then
        '    L.Add(New RuleViolation("Name Required", "Name"))
        'End If
        Return L
    End Function
End Class

File: CombatTracker/CombatTracker.Entities/Reference/Creatures/Charts/LevelChart.cs

using System.Collections.Generic;
using System.Linq;

namespace CombatTracker.Entities.Reference.Creatures.Charts
{
    public partial class LevelChart
    {
        public LevelChart()
        {
            this.LevelChartValues = new HashSet<LevelChartValue>();
            //this.Creatures = new HashSet<Creature>();
        }
    
        public int ID { get; set; }
        public string Name { get; set; }
    
        public virtual ICollection<LevelChartValue> LevelChartValues { get; set; }


        public override string ToString()
        {
            return Name;
        }
        

        public int LookupLevel(int roll)
        {
            return (from lc in LevelChartValues where lc.Min <= roll && roll <= lc.Max select lc.Mod).FirstOrDefault();
        }
    }
    
}
